Publications Office of the European Union.js

```diff
@@ -0,0 +1,222 @@
+{
+	"translatorID": "7d8e6337-3f52-4e8c-8915-95a2ec755b6c",
+	"label": "Publications Office of the European Union",
+	"creator": "Abe Jellinek",
+	"target": "^https?://op\\.europa\\.eu/",
+	"minVersion": "3.0",
+	"maxVersion": "",
+	"priority": 100,
+	"inRepository": true,
+	"translatorType": 4,
+	"browserSupport": "gcsibv",
+	"lastUpdated": "2021-05-24 20:12:14"
+}
+
+/*
+	***** BEGIN LICENSE BLOCK *****
+
+	Copyright © 2021 Abe Jellinek
+	
+	This file is part of Zotero.
+
+	Zotero is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	Zotero is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero. If not, see <http://www.gnu.org/licenses/>.
+
+	***** END LICENSE BLOCK *****
+*/
+
+
+function detectWeb(doc, url) {
+	if (url.includes('/publication-detail/')
+		&& (doc.querySelector('.col-isbn') || doc.querySelector('.col-doi'))) {
+		return "book";
+	}
+	else if (getSearchResults(doc, true)) {
+		return "multiple";
+	}
+	return false;
+}
+
+function getSearchResults(doc, checkOnly) {
+	var items = {};
+	var found = false;
+	var rows = doc.querySelectorAll('.search-results-items a.documentDetailLink');
+	for (let row of rows) {
+		let href = row.href;
+		let title = ZU.trimInternal(row.textContent);
+		if (!href || !title || !href.includes('/publication-detail/')) continue;
+		if (checkOnly) return true;
+		found = true;
+		items[href] = title;
+	}
+	return found ? items : false;
+}
+
+function doWeb(doc, url) {
+	if (detectWeb(doc, url) == "multiple") {
+		Zotero.selectItems(getSearchResults(doc, false), function (items) {
+			if (items) ZU.processDocuments(Object.keys(items), scrape);
+		});
+	}
+	else {
+		scrape(doc, url);
+	}
+}
+
+function scrape(doc, url) {
+	// we would ideally just grab the "Metadata RDF" file here, but it doesn't
+	// give us RDF that Zotero can do anything with. the official metadata
+	// service (opac.publications.europa.eu) doesn't seem especially reliable.
+	// so we'll instead try ISBN and DOI.
+	
+	let pdfLinks = doc.querySelectorAll('a.download');
+	let search = Zotero.loadTranslator("search");
+	
+	// ISBN gives the best data, but we can fall back on DOI if needed
```

Comment:
What are the differences? Generally speaking, we always want to use a DOI over an ISBN, because a DOI offers canonical metadata from the publisher available via stable APIs, whereas ISBN metadata varies and we often have to scrape websites to retrieve it.

Reply:
For this catalog, DOI returns incorrect/very incomplete author lists and almost every field has some weird formatting (extraneous punctuation, etc). ISBN results are all spot on.

Reply:
Well, it seems like the main difference is that they don't register personal authors with the DOI, so the first test case only has the institution. The actual presentation of the institutional author is more correct in the DOI, though — e.g., `European Commission. Representation in Luxemburg.` (correct-ish) vs. "European Commission" and "Representation in Luxemburg" as separate authors (less correct, arguably). But if we think these are cited by the individual authors, including those makes sense.

DOI doesn't have abstract (as is generally the case with DOIs) or ISBN, but it does have `Place: LU` and `Publisher: Publications Office`.

In a case like this, where there's a lot of uniformity, it's probably worth determining how these things should generally be cited and hard-coding values/transformations. E.g., if "Publications Office of the European Union" should just always be in a citation, we could just put that in the Publisher field. Same with the institutional author, which may be limited to a relatively small number of options and could be normalized (and maybe that value makes more sense as a Publisher anyway when there are individual authors?). While we can't exactly be faulted for just saving the metadata they provide, whenever we have a chance to save something such that every single Zotero user who cites it doesn't have to tweak the metadata, we should take the opportunity to do so.